### PR TITLE
Implement API for ad viewability.

### LINKED ISSUE #720: Implement IntersectionObserver polyfill for 3p iframe
Spec is here
http://rawgit.com/slightlyoff/IntersectionObserver/master/index.html
More info: https://github.com/slightlyoff/IntersectionObserver

- The data would come from the parent iframe / AMP resource manager.
- Only polyfill data we have easily available
- Need to decide poll interval
- Need to decide whether to actually polyfill OR just make a very similar API.
- Probably the latter, exposed through a single object (no exposed constructor) in the frame: `context.intersectionObserver`

### PR DESCRIPTION
**Pull request by cramforce**
Exposes `window.context.observeIntersection` to ads that provides largely IntersectionObserver compatible change records of the ad's position in the container viewport.

Also makes the following advancements:
- 3p frames now check that the claimed origin of the parent is correct. This helps ads trust the parent.
- Adds tests for integration.js
- adds `3p-frame/listenOnce,postMessage`
- adds `rectIntersection` to layout-rect.js

Fixes #720

Filed #873 to track increasing fidelity of change records.

### CODE REVIEW

File: test/_init_tests.js [hunk 1]
```diff
@@ -129,3 +129,17 @@ chai.Assertion.addMethod('display', function(display) {
     value
   );
 });
+
+chai.Assertion.addMethod('jsonEqual', function(compare) {
```
Comment (erwinmombay) [inline test/_init_tests.js]: nice :+1:

File: src/3p-frame.js [hunk 2]
```diff
@@ -138,6 +139,35 @@ export function listen(iframe, typeOfMessage, callback) {
...
 /**
+ * Allows listening for a message from the iframe and then removes the listener
+ *
+ * @param {!Element} iframe
+ * @param {string} typeOfMessage
+ * @param {function(!Object)} callback Called when a message of this type
+ *     arrives for this iframe.
+ * @return {!Unlisten}
+ */
+export function listenOnce(iframe, typeOfMessage, callback) {
```
Comment (erwinmombay) [inline src/3p-frame.js]: closes https://github.com/ampproject/amphtml/issues/861

**Comment by cramforce** [comment]
CC @dvoytenko

**Comment by cramforce** [comment]
@erwinmombay Please take a look.

**Comment by erwinmombay** [comment]
@cramforce yep, sorry reading through it as we type this :)

**Comment by cramforce** [comment]
@erwinmombay Cool, just wanted to make sure you saw it. Sorry for the large change!

File: 3p/integration.js [hunk 3]
```diff
@@ -113,6 +120,51 @@ function nonSensitiveDataPostMessage(type, opt_object) {
...
+ * @visibleForTesting
+ */
+export function validateParentOrigin(window, parentLocation) {
+  var ancestors = window.location.ancestorOrigins;
+  if (ancestors == null || !ancestors.length) {
+    parentLocation.originValidated = false;
+    return;
+  }
+  if (ancestors[0] != parentLocation.origin) {
+    throw new Error('Parent origin mismatch: ' +
```
Comment (erwinmombay) [inline 3p/integration.js]: can we use `assert` here instead? (since we already import it)

File: ads/README.md [hunk 4]
```diff
@@ -29,13 +29,32 @@ We will provide the following information to the ad:
 
 - `window.context.referrer` contains the origin of the referrer value of the primary document if available.
 - `document.referrer` will typically contain the URL of the primary document. This may change in the future (See next value for a more reliable method).
-- `window.context.location.href` contains the sanitized location.href value of the primary document.
+- `window.context.location` contains the sanitized Location object of the primary document.
```
Comment (erwinmombay) [inline ads/README.md]: `Location`

File: ads/README.md [hunk 5]
```diff
@@ -29,13 +29,32 @@ We will provide the following information to the ad:
...
 - `window.context.canonicalUrl` contains the canonical URL of the primary document as defined by its `link rel=canonical` tag.
-- ad viewability: Tracked in https://github.com/ampproject/amphtml/issues/720
+- [ad viewability](#ad-viewability)
 - `window.context.noContentAvailable` is a function that the ad system can call if the ad slot was not filled. The container page will then react by showing placeholder content (not by collapsing the ad slot; sizing rules apply).
 
 More information can be provided in a similar fashion if needed (Please file an issue).
 
+### Ad viewability
+
+Ads can all the special API `window.context.observeIntersection(changesCallback)` to receive IntersectionObserver style [change records](http://rawgit.com/slightlyoff/IntersectionObserver/master/index.html#intersectionobserverentry) of the ad's intersection with the parent viewport.
```
Comment (erwinmombay) [inline ads/README.md]: call

File: ads/README.md [hunk 6]
```diff
@@ -29,13 +29,32 @@ We will provide the following information to the ad:
...
+
+The API allows specifying a callback that fires with change records when AMP observes that an ad becomes visible and then while it is visible changes are reported as they happen.
```
Comment (erwinmombay) [inline ads/README.md]: add a comma after `while it is visible`, to make flow a bit better.

...and then while it is visible, changes are reported as they happen.

File: src/layout-rect.js [hunk 7]
```diff
@@ -83,6 +83,28 @@ export function layoutRectsOverlap(r1, r2) {
...
+export function rectIntersection(a, b) {
+  const x0 = Math.max(a.left, b.left);
+  const x1 = Math.min(a.left + a.width, b.left + b.width);
+
+  if (x0 <= x1) {
+    const y0 = Math.max(a.top, b.top);
+    const y1 = Math.min(a.top + a.height, b.top + b.height);
+
+    if (y0 <= y1) {
+      return new layoutRectLtwh(x0, y0, x1 - x0, y1 - y0);
```
Comment (erwinmombay) [inline src/layout-rect.js]: is `layoutRectLtwh` a constructor/class? why do we need `new` here.

File: src/vsync.js [hunk 8]
```diff
@@ -87,24 +87,35 @@ export class Vsync {
...
+        tasks[i].measure(states[i]);
       }
-      for (let i = 0; i < tasks.length; i++) {
-        if (tasks[i].mutate) {
-          tasks[i].mutate(states[i]);
-        }
+    }
+    for (let i = 0; i < tasks.length; i++) {
+      if (tasks[i].mutate) {
+        tasks[i].mutate(states[i]);
```
Comment (erwinmombay) [inline src/vsync.js]: nice change!

File: 3p/integration.js [hunk 9]
```diff
@@ -113,6 +120,51 @@ function nonSensitiveDataPostMessage(type, opt_object) {
...
+ * Throws if the current frame's parent origin is not equal to
+ * the claimed origin.
+ * For browsers that don't support ancestorOrigins it adds
+ * `originValidated = false` to the location object.
+ * @param {!Window} window
+ * @param {!Location} parentLocation
+ * @visibleForTesting
+ */
+export function validateParentOrigin(window, parentLocation) {
+  var ancestors = window.location.ancestorOrigins;
```
Comment (erwinmombay) [inline 3p/integration.js]: think we can `const` here.

File: src/3p.js [hunk 10]
```diff
@@ -26,13 +26,17 @@ import {assert} from './asserts';
...
 
 /** @typedef {function(!Window, !Object)}  */
-var ThirdPartyFunction;
+let ThirdPartyFunction;
 
 
-/** @const {!Object<ThirdPartyFunction>} */
-var registrations = {};
+/**
+ * @const {!Object<ThirdPartyFunction}
```
Comment (erwinmombay) [inline src/3p.js]: missing `>`

**Comment by erwinmombay** [comment]
handing off to @dvoytenko in case he has anything to add

File: test/functional/test-layout-rect.js [hunk 11]
```diff
@@ -69,4 +69,20 @@ describe('LayoutRect', () => {
...
+    const rect1 = lr.layoutRectLtwh(10, 20, 40, 50);
+    const rect2 = lr.layoutRectLtwh(40, 60, 10, 10);
+    const rect3 = lr.layoutRectLtwh(1000, 60, 10, 10);
+    expect(lr.rectIntersection(rect1, rect2)).to.jsonEqual({
+      "left": 40,
+      "top": 60,
+      "width": 10,
+      "height": 10,
+      "bottom": 70,
+      "right":50
```
Comment (erwinmombay) [inline test/functional/test-layout-rect.js]: space after `:`

File: 3p/integration.js [hunk 12]
```diff
@@ -113,6 +120,51 @@ function nonSensitiveDataPostMessage(type, opt_object) {
...
+  if (ancestors == null || !ancestors.length) {
+    parentLocation.originValidated = false;
```
Comment (dvoytenko) [inline 3p/integration.js]: So, this is off for the browsers that don't have ancestorOrigins, right? If so, it'd probably be cleaner to check of `!ancestors` given that `undefined` is the real value for us here.

File: builtins/amp-ad.js [hunk 13]
```diff
@@ -244,14 +278,72 @@ export function installAd(win) {
...
+    viewportCallback(inViewport) {
+      // Lets the ad know that it became visible or no longer is.
+      this.sendAdIntersection_();
+      // And update the ad about its position in the viewport while
+      // it is visible.
+      if (inViewport) {
+        this.unlistenViewportChanges_ =
+            this.getViewport().onChanged(this.sendAdIntersection_.bind(this));
+      } else {
+        this.unlistenViewportChanges_();
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: While unlikely, but let's check on null value for `this.unlistenViewportChanges_` and null it after calling it.

File: src/intersection-observer.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Produces a change entry for that should be compatible with
+ * IntersectionObserverEntry.
+ *
+ * Mutates passed in rootBounds to have x and y according to spec.
+ *
+ * @param {number} time Time when values below were measured.
+ * @param {!LayoutRect} rootBounds Equivalent to viewport.getRect()
+ * @param {!LayoutRect} iframeLayoutBox Layout box of the element
```
Comment (dvoytenko) [inline src/intersection-observer.js]: elementLayoutBox? Sounds like this is not only about iframe....

File: src/viewport.js [hunk 15]
```diff
@@ -202,9 +202,9 @@ export class Viewport {
    * @return {!LayoutRect}
    */
   getRect() {
-    var scrollTop = this.binding_.getScrollTop();
-    var scrollLeft = this.binding_.getScrollLeft();
-    var size = this.binding_.getSize();
+    var scrollTop = this.getScrollTop();
```
Comment (dvoytenko) [inline src/viewport.js]: `const`

File: src/vsync.js [hunk 16]
```diff
@@ -116,6 +127,14 @@ export class Vsync {
   }
 
   /**
+   * Runs the measure operation via vsync.
+   * @param {function()} mutator
```
Comment (dvoytenko) [inline src/vsync.js]: Rename to `measurer`?

File: 3p/integration.js (hunk 9 above)
Comment (cramforce) [inline 3p/integration.js]: Done.

File: 3p/integration.js (hunk 12 above)
Comment (cramforce) [inline 3p/integration.js]: Done

File: 3p/integration.js (hunk 3 above)
Comment (cramforce) [inline 3p/integration.js]: Done.

File: ads/README.md (hunk 4 above)
Comment (cramforce) [inline ads/README.md]: Done

File: ads/README.md (hunk 5 above)
Comment (cramforce) [inline ads/README.md]: Done

File: ads/README.md (hunk 6 above)
Comment (cramforce) [inline ads/README.md]: Done

File: builtins/amp-ad.js (hunk 13 above)
Comment (cramforce) [inline builtins/amp-ad.js]: Done

File: src/3p.js (hunk 10 above)
Comment (cramforce) [inline src/3p.js]: Done

File: src/intersection-observer.js (hunk 14 above)
Comment (cramforce) [inline src/intersection-observer.js]: Yep, missed in refactoring.

File: src/layout-rect.js (hunk 7 above)
Comment (cramforce) [inline src/layout-rect.js]: Good catch.

File: src/viewport.js (hunk 15 above)
Comment (cramforce) [inline src/viewport.js]: Done

File: src/vsync.js (hunk 16 above)
Comment (cramforce) [inline src/vsync.js]: Why. The equivalent API is calle `mutate`

File: test/functional/test-layout-rect.js (hunk 11 above)
Comment (cramforce) [inline test/functional/test-layout-rect.js]: Done.

File: src/vsync.js (hunk 16 above)
Comment (dvoytenko) [inline src/vsync.js]: I mean the parameter "mutator". It should be renamed to make clear that it's for measuring.

**Comment by dvoytenko** [comment]
LGTM with a small comment

**Comment by erwinmombay** [comment]
LGTM

**Pull request merged**